File: gnd/src/main/java/com/google/android/gnd/ui/common/ViewModelModule.java
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.ui.common;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import com.google.android.gnd.MainViewModel;
import com.google.android.gnd.ui.datacollection.DataCollectionViewModel;
import com.google.android.gnd.ui.editsubmission.DateFieldViewModel;
import com.google.android.gnd.ui.editsubmission.EditSubmissionViewModel;
import com.google.android.gnd.ui.editsubmission.MultipleChoiceFieldViewModel;
import com.google.android.gnd.ui.editsubmission.NumberFieldViewModel;
import com.google.android.gnd.ui.editsubmission.PhotoFieldViewModel;
import com.google.android.gnd.ui.editsubmission.TextFieldViewModel;
import com.google.android.gnd.ui.editsubmission.TimeFieldViewModel;
import com.google.android.gnd.ui.home.HomeScreenViewModel;
import com.google.android.gnd.ui.home.featuredetails.FeatureDetailsViewModel;
import com.google.android.gnd.ui.home.featuredetails.SubmissionListItemViewModel;
import com.google.android.gnd.ui.home.featuredetails.SubmissionListViewModel;
import com.google.android.gnd.ui.home.featureselector.FeatureSelectorViewModel;
import com.google.android.gnd.ui.home.mapcontainer.FeatureRepositionViewModel;
import com.google.android.gnd.ui.home.mapcontainer.MapContainerViewModel;
import com.google.android.gnd.ui.home.mapcontainer.PolygonDrawingViewModel;
import com.google.android.gnd.ui.offlinebasemap.OfflineAreasViewModel;
import com.google.android.gnd.ui.offlinebasemap.selector.OfflineAreaSelectorViewModel;
import com.google.android.gnd.ui.offlinebasemap.viewer.OfflineAreaViewerViewModel;
import com.google.android.gnd.ui.signin.SignInViewModel;
import com.google.android.gnd.ui.submissiondetails.SubmissionDetailsViewModel;
import com.google.android.gnd.ui.surveyselector.SurveySelectorViewModel;
import com.google.android.gnd.ui.syncstatus.SyncStatusViewModel;
import com.google.android.gnd.ui.tos.TermsOfServiceViewModel;
import dagger.Binds;
import dagger.Module;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import dagger.multibindings.IntoMap;

@InstallIn(SingletonComponent.class)
@Module
public abstract class ViewModelModule {

  @Binds
  @IntoMap
  @ViewModelKey(FeatureRepositionViewModel.class)
  abstract ViewModel bindFeatureRepositionViewModel(FeatureRepositionViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(PolygonDrawingViewModel.class)
  abstract ViewModel bindPolygonDrawingViewModel(PolygonDrawingViewModel viewModel);


  @Binds
  @IntoMap
  @ViewModelKey(MapContainerViewModel.class)
  abstract ViewModel bindMapContainerViewModel(MapContainerViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreaSelectorViewModel.class)
  abstract ViewModel bindOfflineAreaSelectorViewModel(OfflineAreaSelectorViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SyncStatusViewModel.class)
  abstract ViewModel bindSyncStatusViewModel(SyncStatusViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(DataCollectionViewModel.class)
  abstract ViewModel bindDataCollectionViewModel(DataCollectionViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreasViewModel.class)
  abstract ViewModel bindOfflineAreasViewModel(OfflineAreasViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreaViewerViewModel.class)
  abstract ViewModel bindOfflineAreaViewerViewModel(OfflineAreaViewerViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(MainViewModel.class)
  abstract ViewModel bindMainViewModel(MainViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SignInViewModel.class)
  abstract ViewModel bindSignInVideModel(SignInViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(TermsOfServiceViewModel.class)
  abstract ViewModel bindTermsViewModel(TermsOfServiceViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(HomeScreenViewModel.class)
  abstract ViewModel bindHomeScreenViewModel(HomeScreenViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(FeatureDetailsViewModel.class)
  abstract ViewModel bindFeatureDetailsViewModel(FeatureDetailsViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SurveySelectorViewModel.class)
  abstract ViewModel bindSurveySelectorViewModel(SurveySelectorViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SubmissionListItemViewModel.class)
  abstract ViewModel bindSubmissionListItemViewModel(SubmissionListItemViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SubmissionListViewModel.class)
  abstract ViewModel bindSubmissionListViewModel(SubmissionListViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(SubmissionDetailsViewModel.class)
  abstract ViewModel bindSubmissionDetailsViewModel(SubmissionDetailsViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(EditSubmissionViewModel.class)
  abstract ViewModel bindEditSubmissionViewModel(EditSubmissionViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(PhotoFieldViewModel.class)
  abstract ViewModel bindPhotoFieldViewModel(PhotoFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(MultipleChoiceFieldViewModel.class)
  abstract ViewModel bindMultipleChoiceFieldViewModel(MultipleChoiceFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(TextFieldViewModel.class)
  abstract ViewModel bindTextFieldViewModel(TextFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(NumberFieldViewModel.class)
  abstract ViewModel bindNumberFieldViewModel(NumberFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(DateFieldViewModel.class)
  abstract ViewModel bindDateFieldViewModel(DateFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(TimeFieldViewModel.class)
  abstract ViewModel bindTimeFieldViewModel(TimeFieldViewModel viewModel);

  @Binds
  @IntoMap
  @ViewModelKey(FeatureSelectorViewModel.class)
  abstract ViewModel bindFeatureSelectorViewModel(FeatureSelectorViewModel viewModel);

  @Binds
  abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);
}

File: app/src/main/java/com/google/android/ground/ui/common/ViewModelModule.kt
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.ground.ui.common

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.google.android.ground.MainViewModel
import com.google.android.ground.ui.datacollection.tasks.date.DateTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.location.CaptureLocationTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.multiplechoice.MultipleChoiceTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.number.NumberTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.photo.PhotoTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.point.DropPinTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.polygon.DrawAreaTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.text.TextTaskViewModel
import com.google.android.ground.ui.datacollection.tasks.time.TimeTaskViewModel
import com.google.android.ground.ui.home.HomeScreenViewModel
import com.google.android.ground.ui.home.mapcontainer.HomeScreenMapContainerViewModel
import com.google.android.ground.ui.home.mapcontainer.MapTypeViewModel
import com.google.android.ground.ui.offlineareas.OfflineAreasViewModel
import com.google.android.ground.ui.offlineareas.selector.OfflineAreaSelectorViewModel
import com.google.android.ground.ui.offlineareas.viewer.OfflineAreaViewerViewModel
import com.google.android.ground.ui.signin.SignInViewModel
import com.google.android.ground.ui.startup.StartupViewModel
import com.google.android.ground.ui.surveyselector.SurveySelectorViewModel
import com.google.android.ground.ui.syncstatus.SyncStatusViewModel
import com.google.android.ground.ui.tos.TermsOfServiceViewModel
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoMap

@InstallIn(SingletonComponent::class)
@Module
abstract class ViewModelModule {
  @Binds
  @IntoMap
  @ViewModelKey(DrawAreaTaskViewModel::class)
  abstract fun bindDrawAreaTaskViewModel(viewModel: DrawAreaTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(HomeScreenMapContainerViewModel::class)
  abstract fun bindMapContainerViewModel(viewModel: HomeScreenMapContainerViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreaSelectorViewModel::class)
  abstract fun bindOfflineAreaSelectorViewModel(viewModel: OfflineAreaSelectorViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(SyncStatusViewModel::class)
  abstract fun bindSyncStatusViewModel(viewModel: SyncStatusViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreasViewModel::class)
  abstract fun bindOfflineAreasViewModel(viewModel: OfflineAreasViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(OfflineAreaViewerViewModel::class)
  abstract fun bindOfflineAreaViewerViewModel(viewModel: OfflineAreaViewerViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(MainViewModel::class)
  abstract fun bindMainViewModel(viewModel: MainViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(SignInViewModel::class)
  abstract fun bindSignInVideModel(viewModel: SignInViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(TermsOfServiceViewModel::class)
  abstract fun bindTermsViewModel(viewModel: TermsOfServiceViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(HomeScreenViewModel::class)
  abstract fun bindHomeScreenViewModel(viewModel: HomeScreenViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(SurveySelectorViewModel::class)
  abstract fun bindSurveySelectorViewModel(viewModel: SurveySelectorViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(PhotoTaskViewModel::class)
  abstract fun bindPhotoTaskViewModel(viewModel: PhotoTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(MultipleChoiceTaskViewModel::class)
  abstract fun bindMultipleChoiceTaskViewModel(viewModel: MultipleChoiceTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(TextTaskViewModel::class)
  abstract fun bindTextTaskViewModel(viewModel: TextTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(NumberTaskViewModel::class)
  abstract fun bindNumberTaskViewModel(viewModel: NumberTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(DateTaskViewModel::class)
  abstract fun bindDateTaskViewModel(viewModel: DateTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(TimeTaskViewModel::class)
  abstract fun bindTimeTaskViewModel(viewModel: TimeTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(DropPinTaskViewModel::class)
  abstract fun bindDropPinTaskViewModel(viewModel: DropPinTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(CaptureLocationTaskViewModel::class)
  abstract fun bindCaptureLocationTaskViewModel(viewModel: CaptureLocationTaskViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(BaseMapViewModel::class)
  abstract fun bindBaseMapViewModel(viewModel: BaseMapViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(MapTypeViewModel::class)
  abstract fun bindMapTypeViewModel(viewModel: MapTypeViewModel): ViewModel

  @Binds
  @IntoMap
  @ViewModelKey(StartupViewModel::class)
  abstract fun bindStartupViewModel(viewModel: StartupViewModel): ViewModel

  @Binds abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory
}
